android/java/org/chromium/chrome/browser/crypto_wallet/util/BalanceHelper.java

```diff
@@ -169,10 +177,77 @@ public static void getBlockchainTokensBalances(JsonRpcService jsonRpcService,
                 }
             }
 
-            callback.call(blockchainTokensBalances);
+            callback.call(selectedNetwork.coin, blockchainTokensBalances);
         });
     }
 
+    public static void getP3ABalances(
+            BraveWalletBaseActivity activity, NetworkInfo selectedNetwork,
+            Callbacks.Callback1<HashMap<Integer, HashSet<String>>> callback) {
+        KeyringService keyringService = activity.getKeyringService();
+        BraveWalletService braveWalletService = activity.getBraveWalletService();
+        BlockchainRegistry blockchainRegistry = activity.getBlockchainRegistry();
+        JsonRpcService jsonRpcService = activity.getJsonRpcService();
+        assert braveWalletService != null && blockchainRegistry != null && keyringService != null  && jsonRpcService != null;
+
+        boolean P3AEnabled = BraveConfig.P3A_ENABLED && BravePrefServiceBridge.getInstance().getP3AEnabled();
+
+        HashMap<Integer, HashSet<String>> activeAddresses = new HashMap<Integer, HashSet<String>>();
+        for (int coinType : Utils.P3ACoinTypes) activeAddresses.put(coinType, new HashSet<String>());
+
+        if (!P3AEnabled) {
+                callback.call(activeAddresses);
+                return;
+        } else {
+                Utils.getP3ANetworks(relevantNetworks -> {
+                        // Exclude selectedNetwork if in relevantNetworks, also sort by CoinType
```

Comment:
Can you remind me why we exclude the selected network?

Reply:
It is to avoid querying the same network twice, the selected one is updated at https://github.com/brave/brave-core/pull/16242/files#diff-30eae7637d039986bb60e7d56d19a411894f2bce4725def3d4dd0945ea5edc72R1525.